File: files/CSEC 08.12.py
import pandas as pd

path = "files/Excel Challenge Dec 8th.xlsx"
input = pd.read_excel(path, usecols="B:D", skiprows=1, nrows=10)
test = pd.read_excel(path, usecols="F:G", skiprows=1, nrows=4)
test.columns = ["Airline", "Whole Route"]

pivot_wider = input.pivot(index='Airline', columns='Numb', values='Route')
pivot_longer = pd.melt(pivot_wider.reset_index(), id_vars=['Airline'], value_vars=pivot_wider.columns, var_name='Numb', value_name='Route')
pivot_longer = pivot_longer.sort_values(['Airline', 'Numb'])
pivot_longer = pivot_longer.dropna()

pivot_longer['Route'] = pivot_longer.apply(lambda row: row['Route'] if row['Numb'] == 1 else (row['Route'].split('-')[1] if len(row['Route'].split('-')) > 1 else row['Route']), axis=1)

result = pivot_longer.groupby('Airline')['Route'].apply(lambda x: ' -'.join(x)).reset_index()
result.columns = ["Airline", "Whole Route"]

print(result.equals(test)) # True

File: files/2024-12-08/CSEC 08.12.R
library(tidyverse)
library(readxl)

path = "files/Excel Challenge Dec 8th.xlsx"
input = read_excel(path, range = "B2:D12")
test  = read_excel(path, range = "F2:G6")

result = input %>%
  pivot_wider(names_from = "Numb", values_from = "Route") %>%                                                                                                                                                                                                                                                                                                                  pivot_wider(names_from = "Numb", values_from = "Route") %>%
  pivot_longer(cols = c(2:5), names_to = "Numb", values_to = "Route") %>%
  filter(!is.na(Route)) %>%
  mutate(Route = ifelse(Numb == 1, Route, str_extract(Route, "\\- [A-Z]+"))) %>%
  summarise(`Whole Route` = paste0(Route, collapse = " "), .by = Airline) %>%
  arrange(Airline)

all.equal(result, test)
#> [1] TRUE